rent/src/lib.rs

```diff
@@ -0,0 +1,1489 @@
+			Ok(().into())
+		}
+
+		/// Renew a rent contract for a subscription period.
+		#[pallet::weight((
+			{
+				let s = Contracts::<T>::get(nft_id)
+					.map_or_else(||0, |c| {
+						let mut queues = Queues::<T>::get();
+						match c.duration {
+							Duration::Subscription(_, _) => queues.subscription_queue.size(),
+							Duration::Fixed(_) => 0,
+							Duration::Infinite => 0
+						}
+					});
+				T::WeightInfo::renew_contract(s as u32)
+			},
+			DispatchClass::Normal
+		))]
+		pub fn renew_contract(origin: OriginFor<T>, nft_id: NFTId) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			let contract = Contracts::<T>::get(nft_id).ok_or(Error::<T>::ContractNotFound)?;
+			ensure!(contract.has_started, Error::<T>::ContractHasNotStarted);
+			let is_subscription = matches!(contract.duration, Duration::Subscription { .. });
+			ensure!(is_subscription, Error::<T>::RenewalOnlyForSubscription);
+			if let Some(rentee) = contract.rentee {
+				Self::take_rent_fee(&rentee, contract.renter.clone(), &contract.rent_fee)?;
+			}
+			// Deposit event.
+			let event = Event::ContractSubscriptionPeriodStarted { nft_id };
+			Self::deposit_event(event);
+
+			Ok(().into())
+		}
+
+		/// Remove a contract from available list if expiration has been reached.
+		#[pallet::weight((
+			{
+				let mut queues = Queues::<T>::get();
+				let s = queues.available_queue.size();
+				T::WeightInfo::remove_expired_contract(s as u32)
+			},
+			DispatchClass::Normal
+		))]
+		pub fn remove_expired_contract(
+			origin: OriginFor<T>,
+			nft_id: NFTId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			let contract = Contracts::<T>::get(nft_id).ok_or(Error::<T>::ContractNotFound)?;
+			let mut nft = T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFound)?;
+
+			Self::process_cancellation_fees(nft_id, &contract, None)?;
+
+			nft.state.is_rented = false;
+			T::NFTExt::set_nft(nft_id, nft)?;
+
+			Offers::<T>::remove(nft_id);
+
+			Contracts::<T>::remove(nft_id);
+
+			// Deposit event.
+			let event = Event::ContractAvailableExpired { nft_id };
+			Self::deposit_event(event);
+
+			Ok(().into())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// The account ID of the cancellation fee pot.
+	pub fn account_id() -> T::AccountId {
+		T::PalletId::get().into_account_truncating()
+	}
+
+	/// Check that an NFT is available for rent.
+	pub fn ensure_nft_available(
+		nft: &NFTData<T::AccountId, <T::NFTExt as NFTExt>::NFTOffchainDataLimit>,
+	) -> Result<(), DispatchError> {
+		ensure!(!nft.state.is_listed, Error::<T>::CannotUseListedNFTs);
+		ensure!(!nft.state.is_capsule, Error::<T>::CannotUseCapsuleNFTs);
+		ensure!(!nft.state.is_delegated, Error::<T>::CannotUseDelegatedNFTs);
+		ensure!(!nft.state.is_soulbound, Error::<T>::CannotUseSoulboundNFTs);
+		ensure!(!nft.state.is_rented, Error::<T>::CannotUseRentedNFTs);
+		Ok(())
+	}
+
+	/// Check that address has NFT / Balance to cover cancellation fee.
+	pub fn ensure_enough_for_cancellation_fee(
```

Comment:
Once you implement the `get_balance` and `get_nft` for CancelationFee you can do then the following:
```rust
	/// Check that address has NFT / Balance to cover cancellation fee.
	pub fn ensure_enough_for_cancellation_fee(
		cancellation_fee: &Option<CancellationFee<BalanceOf<T>>>,
		from: &T::AccountId,
	) -> Result<(), DispatchError> {
		let cancellation_fee = match cancellation_fee {
			Some(x) => x,
			None => return Ok(()),
		};

		if let Some(amount) = cancellation_fee.get_balance() {
			let free_balance = T::Currency::free_balance(from);
			ensure!(free_balance > amount, Error::<T>::NotEnoughBalanceForCancellationFee);
		}

		if let Some(nft_id) = cancellation_fee.get_nft() {
			let nft =
				T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFoundForCancellationFee)?;
			ensure!(nft.owner == *from, Error::<T>::NotTheNFTOwnerForCancellationFee);
			Self::ensure_nft_available(&nft)?;
		}

		Ok(())
	}
```